packages/core/src/directives/template-map.ts

```diff
@@ -0,0 +1,172 @@
+ */
+const setMapped = (element: Element, name: string, value: unknown, oldValue: unknown): void => {
+  const type = getMapType(name);
+  name = getMapName(name);
+  switch (type) {
+    case MAP_TYPE.LISTENER:
+      if (typeof oldValue === 'function') {
+        element.removeEventListener(name, oldValue as EventListener);
+      }
+      if (typeof value === 'function') {
+        element.addEventListener(name, value as EventListener);
+      }
+      break;
+    case MAP_TYPE.PROPERTY:
+      if (value === undefined) {
+        delete (element as Element & { [name: string]: typeof value })[name];
+      }
+      else {
+        (element as Element & { [name: string]: typeof value })[name] = value;
+      }
+      break;
+    case MAP_TYPE.ATTRIBUTE:
+    default:
+      if (value === null || value === undefined) { // remove with undefined to comply with property
+        element.removeAttribute(name);
+      }
+      else {
+        element.setAttribute(name, String(value));
+      }
+      break;
+  }
+};
+
+class TemplateMapDirective extends Directive {
+  private valueMap: ValueMap = {}; // used to cache values
+
+  constructor (partInfo: PartInfo) {
+    super(partInfo);
+
+    if (partInfo.type !== PartType.ELEMENT) {
+      throw new Error('The `TemplateMap` directive must be used with element');
+    }
+  }
+
+  /**
+   * @inheritDoc
+   * @param part Element part
+   * @param attributeMap Attribute map
+   * @returns noChange
+   */
+  public update (part: ElementPart, [attributeMap]: [TemplateMap]): typeof noChange {
+    const newValueMap: ValueMap = {};
+    const element = part.element;
+
+    for (const name in attributeMap) {
+      const value = attributeMap[name];
+      let scopedValue = value;
+      let oldValue;
+      let oldScopedValue;
+
+      if (this.valueMap[name]) {
+        oldValue = this.valueMap[name].value;
+        oldScopedValue = scopedValue = this.valueMap[name].scopedValue;
+      }
+
+      if (oldValue !== value) {
+        // always assume that if `function` is passed it is an event listener
+        scopedValue = typeof value === 'function' ? value.bind(part.options?.host || element) : value;
+        setMapped(element, name, scopedValue, oldScopedValue);
+      }
+
+      newValueMap[name] = { value, scopedValue };
+      delete this.valueMap[name];
+    }
+
+    // Delete old cached values
+    for (const name in this.valueMap) {
+      setMapped(element, name, undefined, this.valueMap[name].scopedValue);
+    }
+
+    this.valueMap = newValueMap;
+
+    return this.render();
+  }
+
+  /**
+   * @inheritDoc
+   * @returns noChange
+   */
+  public render (): typeof noChange {
+    return noChange;
+  }
+}
+
+/**
+ * A directive to apply attributes to template element
+ * @param template template map with the following syntax:
+ * {
+ *   'label': 'value', // set attribute with attribute value, e.g. `<input label="value">`
+ *   'required': '', // set empty attribute, e.g. `<input required>`
```

Comment:
Should we not be supporting booleans for boolean attributes? I think it would be nicer to use, especially when mapping boolean properties to attributes.